public/app/features/alerting/unified/components/rule-editor/LabelsField.tsx

```diff
@@ -1,18 +1,108 @@
 import { css, cx } from '@emotion/css';
-import React, { FC, useMemo } from 'react';
-import { FieldArrayWithId, useFieldArray, useFormContext } from 'react-hook-form';
+import { flattenDeep, compact } from 'lodash';
+import React, { FC, useCallback, useEffect, useMemo, useState } from 'react';
+import { FieldArrayMethodProps, useFieldArray, useFormContext } from 'react-hook-form';
 
 import { GrafanaTheme, SelectableValue } from '@grafana/data';
-import { Button, Field, InlineLabel, Label, useStyles } from '@grafana/ui';
+import { Stack } from '@grafana/experimental';
+import { Button, Field, InlineLabel, Label, useStyles, Tooltip, Icon, Input, LoadingPlaceholder } from '@grafana/ui';
+import { useDispatch } from 'app/types';
+import { RulerRuleGroupDTO } from 'app/types/unified-alerting-dto';
 
-import { RuleFormValues } from '../../types/rule-form';
-import AlertLabelDropdown, { AlertLabelDropdownProps } from '../AlertLabelDropdown';
+import { useUnifiedAlertingSelector } from '../../hooks/useUnifiedAlertingSelector';
+import { fetchRulerRulesIfNotFetchedYet } from '../../state/actions';
+import { RuleFormType, RuleFormValues } from '../../types/rule-form';
+import AlertLabelDropdown from '../AlertLabelDropdown';
 
 interface Props {
   className?: string;
+  suggest: boolean;
+  dataSourceName?: string | null;
 }
 
-const LabelsField: FC<Props> = ({ className }) => {
+const useGetCustomLabels = (dataSourceName: string): { loading: boolean; labelsByKey: Record<string, string[]> } => {
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    dispatch(fetchRulerRulesIfNotFetchedYet(dataSourceName));
+  }, [dispatch, dataSourceName]);
+
+  const rulerRuleRequests = useUnifiedAlertingSelector((state) => state.rulerRules);
+
+  const rulerRequest = rulerRuleRequests[dataSourceName];
+
+  if (!rulerRequest) {
+    return { loading: false, labelsByKey: {} };
+  }
+
+  if (rulerRequest.loading) {
+    return { loading: true, labelsByKey: {} };
+  }
+
+  const result = rulerRequest.result || {};
+
+  //store all labels in a flat array and remove empty values
+  const labels = compact(
+    flattenDeep(
+      Object.keys(result).map((ruleGroupKey) =>
+        result[ruleGroupKey].map((ruleItem: RulerRuleGroupDTO) => ruleItem.rules.map((item) => item.labels))
+      )
+    )
+  );
+
+  const labelsByKey: Record<string, string[]> = {};
+
+  labels.forEach((label: Record<string, string>) => {
+    Object.entries(label).forEach(([key, value]) => {
+      labelsByKey[key] = [...new Set([...(labelsByKey[key] || []), value])];
+    });
+  });
+
+  return { loading: false, labelsByKey };
```

Comment:
Actually I think we could return directly in here loading from the request :

```
  const rulerRequest = rulerRuleRequests[dataSourceName];

  const result = rulerRequest.result || {};

  //store all labels in a flat array and remove empty values
  const labels = compact(
    flattenDeep(
      Object.keys(result).map((ruleGroupKey) =>
        result[ruleGroupKey].map((ruleItem: RulerRuleGroupDTO) => ruleItem.rules.map((item) => item.labels))
      )
    )
  );

  const labelsByKey: Record<string, string[]> = {};

  labels.forEach((label: Record<string, string>) => {
    Object.entries(label).forEach(([key, value]) => {
      labelsByKey[key] = [...new Set([...(labelsByKey[key] || []), value])];
    });
  });

  return { loading: rulerRequest.loading, labelsByKey: labelsByKey };
```

Reply:
Good idea!!